### ISSUE TITLE
GameGuru MAX - Request - Road "Creator"

### ISSUE THREAD

**Issue by MonkeyFrogStudio** [TheGameCreators/GameGuruRepo#3711]
I really would love to see a means of laying down roads and paths on terrain in GGMAX. I've been playing The Outer Worlds and they have these wonderful roads that traverse all manner of terrain types. Here are a few screen shots for examples:

![ss_9cd1358d6c11d41804fac0e0bc24763cc3226bf5 1920x1080](https://user-images.githubusercontent.com/36735507/190875420-4034a515-77b9-4717-b1af-422c245b9ad9.jpg)

![9282_24_outer-worlds-review-world_full](https://user-images.githubusercontent.com/36735507/190875423-3eb6a05e-3322-4bd5-ae93-20616b6b7f5d.jpg)

![images](https://user-images.githubusercontent.com/36735507/190875426-3194f114-f916-41ea-87ab-98223ba7cb5d.jpg)

![mini_The-outer-Worlds-8-780x405](https://user-images.githubusercontent.com/36735507/190875433-76506fa0-65e2-409e-9ddf-29480103b75e.png)

These roads are not just models laid down piece by piece, but are part of a system that uses models to lay down a "track" on top of the terrain. Several other game engines have road creation tools and you can create custom models that these tools can use. It makes it very easy, and very organic, to lay down roads of all types on terrain. 

Please consider adding something like this for MAX.

**Comment by Lemuripest93** [comment]
We seem to share a lot of the same ideas because this too was on my wish list, lol

I haven't played 'The Outer Worlds' yet but a system for basically building a road would be quite neat, being able to basically change the shape of the road, from straight lines to curves, sharp angles and different intersections, train and city building sims does a lot of systems similar to this.

I could also see potential for procedural road generation here too, generating a massive road system and being allowed to tweak certain aspects of the road, say you want to erase one section to make space for a bridge for instance or maybe a parking lot or something. In any case this is a great request.

**Comment by MonkeyFrogStudio** [comment]
The other game engines use a spline-based system to lay down "track", so to speak. Specially made models follow the track. These models could be anything from roads, to barriers, to walls, and more. Basically anything that could be laid out in lines (straight or curved) over terrain. The models are either "chunks" (like the individual pieces of concrete barriers put up as road blocks) to tiling model pieces to make continuous roads, fencing, etc.

Here's an old video (from 2013) on how the S2 Engine handles this:

https://youtu.be/MzU9XMa9K6c

Other engines have their own ways of doing this or addons that do something similar.

**Comment by Necrym59** [comment]
A spline drawing system itself would be good as anything can be ahered to the line it could be a road overlay or fence object or anything for that matter.

**Comment by Tam51** [comment]
I would like to see this too. I think if would be a great idea to add this to GGmax I also think you could possibly you this system to create rivers with a few tweaks to the system. 

Great idea
THKS

**Comment by Hamada880** [comment]
Quick trick I managed to fix by replacing the terrain texture to concrete texture and flat roads to make roads. The only downfall is the pattern of the texture doesn't works well and you can easily see annoying patterns if you tires to paint road texture like this ![image](https://user-images.githubusercontent.com/81198887/191492213-3c33ddc2-2986-4687-9fd9-1c5a48b8cec5.jpeg)

**Comment by MonkeyFrogStudio** [comment]
Yes, you can replace terrain textures and paint in a road or whatever. But when you do so, you loose a terrain texture to use on the terrain. Even so, that's not what I am after here. The examples posted use 3D objects for roads, etc. And in the video I posted, they even use 3D barriers, which are distributed along the curve. _**The idea here isn't to paint roads and paths. We'll be able to do that when they finally open up the terrain for use with custom materials. The idea here is to easily place 3D objects that will 1) follow a path and 2) conform to the terrain.**_

**Comment by Hamada880** [comment]
> Yes, you can replace terrain textures and paint in a road or whatever. But when you do so, you loose a terrain texture to use on the terrain. Even so, that's not what I am after here. The examples posted use 3D objects for roads, etc. And in the video I posted, they even use 3D barriers, which are distributed along the curve. _**The idea here isn't to paint roads and paths. We'll be able to do that when they finally open up the terrain for use with custom materials. The idea here is to easily place 3D objects that will 1) follow a path and 2) conform to the terrain.**_

Ohh I see now what your aiming for, I thought you were looking for quick hack to paint roads and still use the bpr to add depth.

**Comment by MonkeyFrogStudio** [comment]
Yeah, if you look closely at the examples from The Outer Worlds that I posted, you will see that that center rail in the road is a 3D model that follows a spline path (and the terrain). The road in the top image is also generated from a 3D model. And as was pointed out, such a system doesn't have to be limited to just roads - fences, barriers, etc.

**Comment by Necrym59** [comment]
Another good thing about splines is the capability for animated characters or objects could be made to follow the splines thus creating automated vehicles or carriers etc.

**Comment by MonkeyFrogStudio** [comment]
Here's an example of a real-time river creator for Blender, of all things:

https://youtu.be/L3-9qPpa5DI

The way it uses SPLINES and digs out from the terrain is great.
